src/docbuilder/metadata/lib.rs

```diff
@@ -0,0 +1,631 @@
+//!     cmd.env(key, value);
+//! }
+//!
+//! // Finally, run `cargo doc` on the directory.
+//! let result = cmd.output()?;
+//! # Ok(())
+//! # }
+//! ```
+
+use std::collections::{HashSet, HashMap};
+use std::io;
+use std::path::Path;
+
+use thiserror::Error;
+use toml::{map::Map, Value};
+
+/// The target that `metadata` is being built for.
+///
+/// This is directly passed on from the Cargo [`TARGET`] variable.
+///
+/// [`TARGET`]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts
+pub const HOST_TARGET: &str = env!("DOCS_RS_METADATA_HOST_TARGET");
+/// The targets that are built if no `targets` section is specified.
+///
+/// Currently, this is guaranteed to have only [tier one] targets.
+/// However, it may not contain all tier one targets.
+///
+/// [tier one]: https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-1
+pub const DEFAULT_TARGETS: &[&str] = &[
+    "i686-pc-windows-msvc",
+    "i686-unknown-linux-gnu",
+    "x86_64-apple-darwin",
+    "x86_64-pc-windows-msvc",
+    "x86_64-unknown-linux-gnu",
+];
+
+/// The possible errors for [`Metadata::from_crate_root`].
+#[derive(Debug, Error)]
+pub enum MetadataError {
+    /// The error returned when the manifest could not be read.
+    #[error("failed to read manifest from disk")]
+    IO(#[from] io::Error),
+    /// The error returned when the manifest could not be parsed.
+    #[error("failed to parse manifest")]
+    Parse(#[from] toml::de::Error),
+}
+
+/// Metadata to set for custom builds.
+///
+/// This metadata is read from `[package.metadata.docs.rs]` table in `Cargo.toml`.
+///
+/// An example metadata:
+///
+/// ```text
+/// [package]
+/// name = "test"
+///
+/// [package.metadata.docs.rs]
+/// features = [ "feature1", "feature2" ]
+/// all-features = true
+/// no-default-features = true
+/// default-target = "x86_64-unknown-linux-gnu"
+/// targets = [ "x86_64-apple-darwin", "x86_64-pc-windows-msvc" ]
+/// rustc-args = [ "--example-rustc-arg" ]
+/// rustdoc-args = [ "--example-rustdoc-arg" ]
+/// ```
+///
+/// You can define one or more fields in your `Cargo.toml`.
+pub struct Metadata {
+    /// List of features to pass on to `cargo`.
+    ///
+    /// By default, docs.rs will only build default features.
+    pub features: Option<Vec<String>>,
+
+    /// Whether to pass `--all-features` to `cargo`.
+    pub all_features: bool,
+
+    /// Whether to pass `--no-default-features` to `cargo`.
+    //
+    /// By default, Docs.rs will build default features.
+    /// Set `no-default-fatures` to `false` if you want to build only certain features.
+    pub no_default_features: bool,
+
+    /// See [`BuildTargets`].
+    default_target: Option<String>,
+    targets: Option<Vec<String>>,
+
+    /// List of command line arguments for `rustc`.
+    pub rustc_args: Option<Vec<String>>,
+
+    /// List of command line arguments for `rustdoc`.
+    pub rustdoc_args: Option<Vec<String>>,
```

Comment:
Hmm, why are the fields public?

Reply:
This lets the API user add more flags if they want to. It also makes it possible to construct a `Metadata` without having a TOML file already.

Reply:
You can't construct the metadata since `default_target` and `targets` are not public.

Reply:
`let metadata = Metadata { your_fields, ..Metadata::default() };`

But if you don't think it's useful, I can remove it.

Reply:
Yeah I'd avoid making the fields public.

Reply:
The only downside here is that the documentation for the fields no longer shows up on docs.rs, you have to read the source.

Reply:
Well, the documentation for those fields is in https://docs.rs/about/metadata

Reply:
Ooh good idea, let me link to that.